pyuvdata/uvdata/uvdata.py

```diff
@@ -2360,6 +2380,167 @@ def _check_freq_spacing(self, raise_errors=True):
+                    pol, self.polarization_array, return_indices=True
+                )[2][0]
+                spw_screen = self.flex_spw_id_array == spw
+
+                # Note that this works because pol_idx is an integer, ergo a simple
+                # slice (wherease spw_screen is a complex slice, which we can only have
+                # one of for an array).
+                if self.future_array_shapes:
+                    new_param[:, spw_screen, pol_idx] = param[:, spw_screen, 0]
+                else:
+                    new_param[:, :, spw_screen, pol_idx] = param[:, :, spw_screen, 0]
+
+            # With the new array defined and filled, set the attribute equal to it
+            setattr(self, name, new_param)
+
+        # Finally, remove the flex-pol attribute
+        self.flex_spw_polarization_array = None
+
+    def _make_flex_pol(self, raise_error=False, raise_warning=True):
+        """
+        Convert a regular UVData object into one with flex-polarization enabled.
+
+        This is an internal helper function, which is not designed to be called by
+        users, but rather individual read/write functions for the UVData object.
+        This will convert a regulat UVData object into one that uses flexible
+        polarization, which allows for each spectral window to have its own unique
+        polarization code, useful for storing data more compactly when different
+        windows have different polarizations recorded. Note that at this time,
+        only one polarization code per-spw is allowed -- if more than one polarization
+        is found to have unflagged data in a given spectral window, then the object
+        will not be converted.
+
+        Parameters
+        ----------
+        raise_error : bool
+            If an object cannot be converted to flex-pol, then raise a ValueError.
+            Default is False.
+        raise_warning : bool
+            If an object cannot be converted to flex-pol, and `raise_error=False`, then
+            raise a warning that the conversion failed. Default is True.
+        """
+        if not self.flex_spw:
+            msg = "Cannot make a flex-pol UVData object if flex_spw=False."
+            if raise_error:
+                raise ValueError(msg)
+            if raise_warning:
+                warnings.warn(msg)
+            return
+
+        if self.Npols == 1:
+            # This is basically a no-op, fix the to pol-array attributes and exit
+            if self.flex_spw_polarization_array is None:
+                self.flex_spw_polarization_array = (
+                    np.zeros_like(self.spw_array) + self.polarization_array[0]
+                )
+                self.polarization_array = np.array([0])
+            return
+
+        flex_pol_idx = np.zeros_like(self.spw_array)
+        for idx, spw in enumerate(self.spw_array):
+            spw_screen = self.flex_spw_id_array == spw
+
+            if self.future_array_shapes:
```

Comment:
Might add a comment to say why the results of np.all in both cases have to be flipped with ~

Reply:
Added in 8f8a8af4.